### PR TITLE
Fix error message if NonlinearExpression is mixed with new NL API

### LINKED ISSUE #3740: Should error when trying to mix `NonlinearExpression` and `GenericNonlinearExpr`
Due to a hacky patch on the old version of InfiniteOpt that uses the `NonlinearExpression`, it has come to my attention that the current JuMP API let's users make `GenericNonlinearExpr`s that contain `NonlinearExpression`s which ultimately culminates in hard to diagnose error:
```julia
julia> using JuMP; m = Model(); @variable(m, x);

julia> @NLexpression(m, myexpr, x^3)
subexpression[1]: x ^ 3.0

julia> @expression(m, newexpr, x * myexpr) # produces a NonlinearExpr that contains the NonlinearExpression
x * (subexpression[1]: x ^ 3.0) 

julia> @objective(m, Min, newexpr)
ERROR: MethodError: no method matching check_belongs_to_model(::NonlinearExpression, ::Model)

Closest candidates are:
  check_belongs_to_model(::ConstraintRef, ::AbstractModel)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:63
  check_belongs_to_model(::ScalarConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:614
  check_belongs_to_model(::VectorConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:674
  ...

Stacktrace:
 [1] check_belongs_to_model(expr::NonlinearExpr, model::Model)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\nlp_expr.jl:523
 [2] set_objective_function
   @ ~\.julia\packages\JuMP\as6Ji\src\objective.jl:279 [inlined]
 [3] set_objective(model::Model, sense::MathOptInterface.OptimizationSense, func::NonlinearExpr)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\objective.jl:326
 [4] macro expansion
   @ ~\.julia\packages\JuMP\as6Ji\src\macros\@objective.jl:70 [inlined]
 [5] top-level scope
   @ REPL[64]:1
```
We get a similar error when trying to to set the objective with the `NonlinearExpression` directly:
```julia
julia> @objective(m, Min, myexpr)
ERROR: MethodError: no method matching check_belongs_to_model(::NonlinearExpression, ::Model)

Closest candidates are:
  check_belongs_to_model(::ConstraintRef, ::AbstractModel)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:63
  check_belongs_to_model(::ScalarConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:614
  check_belongs_to_model(::VectorConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\constraints.jl:674
  ...

Stacktrace:
 [1] set_objective_function(model::Model, func::NonlinearExpression)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\objective.jl:279
 [2] set_objective(model::Model, sense::MathOptInterface.OptimizationSense, func::NonlinearExpression)
   @ JuMP ~\.julia\packages\JuMP\as6Ji\src\objective.jl:326
 [3] macro expansion
   @ ~\.julia\packages\JuMP\as6Ji\src\macros\@objective.jl:70 [inlined]
 [4] top-level scope
   @ REPL[65]:1
```
Also, `myexpr` (which is the faulty `NonlinearExpr`) works with `@NLobjective`:
```julia
@NLobjective(m, Min, myexpr)
``` 
It seems like more error checking should be added to JuMP to prevent users from mixing the two nonlinear interfaces. In particular, I think we should do the following:
- Have `check_belongs_to_model` throw a helpful error if a `NonlinearExpression` is given
- Have the `GenericNonlinearExpr` constructor throw an error if it is given a `NonlinearExpression`
- Have the `NL` macros error if they are given `GenericNonlinearExpr`s

### PR DESCRIPTION
**Pull request by odow**
Closes #3740

TODO

 - [x] https://github.com/jump-dev/JuMP.jl/actions/runs/8914238919

### CODE REVIEW

File: test/test_nlp_expr.jl [hunk 1]
```diff
@@ -371,26 +371,6 @@ function test_extension_recursion_stackoverflow(
...
-    @test e isa GenericNonlinearExpr
-    @test string(e) == "x + ($p)"
-    return
-end
-
-function test_nlexpression_interaction()
-    model = Model()
-    @variable(model, x)
-    @NLexpression(model, expr, sin(x))
-    e = x + expr
```
Comment (odow) [inline test/test_nlp_expr.jl:388]: @pulsipher I'm not really sure what I intended to happen here.

Comment (pulsipher) [reply test/test_nlp_expr.jl:388]: I am not sure, but I agree it should be removed in favor of this PR.

**Comment by pulsipher** [comment]
This PR addresses mixing `NonlinearExpression`s with `GenericNonlinearExpr`, but we still get a `MethodError` if a `NonlinearExpression` is passed directly to `@objective` or `@constraint`:
```julia
using JuMP
m = Model()
@variable(m, x)
@NLexpression(m, nl_ex, x^3)
@objective(m, Min, nl_ex)
```
```julia
ERROR: MethodError: no method matching check_belongs_to_model(::NonlinearExpression, ::Model)

Closest candidates are:
  check_belongs_to_model(::ConstraintRef, ::AbstractModel)
   @ JuMP ~\.julia\packages\JuMP\027Gt\src\constraints.jl:63
  check_belongs_to_model(::ScalarConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\027Gt\src\constraints.jl:614
  check_belongs_to_model(::VectorConstraint, ::Any)
   @ JuMP ~\.julia\packages\JuMP\027Gt\src\constraints.jl:674
  ...

Stacktrace:
 [1] set_objective_function(model::Model, func::NonlinearExpression)
   @ JuMP ~\.julia\packages\JuMP\027Gt\src\objective.jl:129
 [2] set_objective(model::Model, sense::MathOptInterface.OptimizationSense, func::NonlinearExpression)
   @ JuMP ~\.julia\packages\JuMP\027Gt\src\objective.jl:176
 [3] macro expansion
   @ ~\.julia\packages\JuMP\027Gt\src\macros\@objective.jl:70 [inlined]
 [4] top-level scope
   @ REPL[5]:1
```
This can be remedied by adding something like:
```julia
function check_belongs_to_model(::NonlinearExpression, ::GenericModel)
   error("Expressions made with `@NLexpression` are only compatible for use with `@NLobjective` and `@NLconstraint`")
end
```

**Comment by odow** [comment]
>  if a NonlinearExpression is passed directly to @objective or @constraint:

Hahah but whhyyyyy would you do this. I'll fix :smile:

**Comment by pulsipher** [comment]
> > if a NonlinearExpression is passed directly to @objective or @constraint:
> 
> Hahah but whhyyyyy would you do this. I'll fix 😄

Never underestimate the power of typos and/or ignorance 😅

**Comment by odow** [comment]
Okay. I've actually tweaked things slightly, because this has nothing to do with the new API, it would also have been a bug in older versions of JuMP.

We now continue to play laissez-faire with expression construction--for better and worse--but we now error when we attempt to convert them to MOI functions. This provides a much more uniform error message for all the possible ways that they could end up being passed to the solver, and we can provide a nice error message.

**Comment by pulsipher** [comment]
I understand the logic behind relying on `moi_function` to provide the appropriate errors, but it still seems quite strange to me to have tests that check that mixing legacy NL objects with `GenericNonlinearExpr` works:
https://github.com/jump-dev/JuMP.jl/blob/fbe96e521bc5ecaacb1e59f57f97df34626f2d4b/test/test_nlp_expr.jl#L374-L438
Why would this be expected behavior that we have tests that ensure it remains? What is the purpose of these tests?

**Comment by odow** [comment]
Okay, how about now.

File: test/test_nlp_expr.jl [hunk 2]
```diff
@@ -1092,4 +1060,55 @@ function test_error_nonlinear_expr_complex_constructor()
...
+
+        To update, replace all calls to `@NLexpression` with `@expression`.
+        """,
+    )
+    @test_throws err @objective(model, Min, arg)
+    @test_throws err @constraint(model, arg <= 0)
+    @test_throws err @constraint(model, arg in MOI.LessThan(0.0))
+    @test_throws err moi_function(arg)
+    @test_throws err x * arg
+    @test_throws err arg * x
```
Comment (odow) [inline test/test_nlp_expr.jl:1081]: There's technically two unrelated bugs here: something like `@objective(model, Min, arg)`, and `x * arg`. But now we uniformly throw the same error, which is hopefully informative for the user.

**Review by pulsipher — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**